mm2src/coins/sia/http_client.rs

```diff
@@ -0,0 +1,193 @@
+pub enum SiaApiClientError {
+    Timeout(String),
+    BuildError(String),
+    ApiUnreachable(String),
+    ReqwestError(ReqwestErrorWithUrl),
+    UrlParse(url::ParseError),
+}
+
+impl From<SiaApiClientError> for String {
+    fn from(e: SiaApiClientError) -> Self { format!("{:?}", e) }
+}
+
+async fn fetch_and_parse<T: DeserializeOwned>(client: &Client, url: Url) -> Result<T, SiaApiClientError> {
+    client
+        .get(url.clone())
+        .send()
+        .await
+        .map_err(|e| {
+            SiaApiClientError::ReqwestError(ReqwestErrorWithUrl {
+                error: e,
+                url: url.clone(),
+            })
+        })?
+        .json::<T>()
+        .await
+        .map_err(|e| SiaApiClientError::ReqwestError(ReqwestErrorWithUrl { error: e, url }))
+}
+
+// https://github.com/SiaFoundation/core/blob/4e46803f702891e7a83a415b7fcd7543b13e715e/types/types.go#L181
+#[derive(Deserialize, Serialize, Debug)]
+pub struct GetConsensusTipResponse {
+    pub height: u64,
+    pub id: String, // TODO this can match "BlockID" type
+}
+
+// https://github.com/SiaFoundation/walletd/blob/9574e69ff0bf84de1235b68e78db2a41d5e27516/api/api.go#L36
+// https://github.com/SiaFoundation/walletd/blob/9574e69ff0bf84de1235b68e78db2a41d5e27516/wallet/wallet.go#L25
+#[derive(Deserialize, Serialize, Debug)]
+pub struct GetAddressesBalanceResponse {
+    pub siacoins: MmNumber,
+    #[serde(rename = "immatureSiacoins")]
+    pub immature_siacoins: MmNumber,
+    pub siafunds: u64,
+}
+
+impl SiaApiClientImpl {
+    fn new(base_url: Url, password: &str) -> Result<Self, SiaApiClientError> {
+        let mut headers = HeaderMap::new();
+        let auth_value = format!("Basic {}", base64::encode(&format!(":{}", password)));
+        headers.insert(
+            AUTHORIZATION,
+            HeaderValue::from_str(&auth_value).map_err(|e| SiaApiClientError::BuildError(e.to_string()))?,
+        );
+
+        let client = reqwest::Client::builder()
+            .default_headers(headers)
+            .timeout(Duration::from_secs(10)) // TODO make this configurable
+            .build()
+            .map_err(|e| {
+                SiaApiClientError::ReqwestError(ReqwestErrorWithUrl {
+                    error: e,
+                    url: base_url.clone(),
+                })
+            })?;
+        Ok(SiaApiClientImpl { client, base_url })
+    }
+
+    pub async fn get_consensus_tip(&self) -> Result<GetConsensusTipResponse, SiaApiClientError> {
+        let base_url = self.base_url.clone();
+        let endpoint_url = base_url
+            .join("api/consensus/tip")
+            .map_err(SiaApiClientError::UrlParse)?;
+
+        fetch_and_parse::<GetConsensusTipResponse>(&self.client, endpoint_url).await
+    }
+
+    pub async fn get_addresses_balance(&self, address: &str) -> Result<GetAddressesBalanceResponse, SiaApiClientError> {
+        let base_url = self.base_url.clone();
+
+        // TODO Validate or sanitize `address` here if necessary
```

Comment:
I don't think we would need this validation since the address will be passed from siacoin struct where it's saved. It would have been validated when created.